packages/block-library/src/gallery/index.php

```diff
@@ -32,6 +32,44 @@ function block_core_gallery_data_id_backcompatibility( $parsed_block ) {
 
 add_filter( 'render_block_data', 'block_core_gallery_data_id_backcompatibility' );
 
+/**
+ * Adds a style tag for the --wp--style--unstable-gallery-gap var.
+ *
+ * The Gallery block needs to recalculate Image block width based on
+ * the current gap setting in order to maintain the number of flex columns
+ * so a css var is added to allow this.
+ *
+ * @param array  $attributes Attributes of the block being rendered.
+ * @param string $content Content of the block being rendered.
+ * @return string The content of the block being rendered.
+ */
+function block_core_gallery_render( $attributes, $content ) {
+	$gap = _wp_array_get( $attributes, array( 'style', 'spacing', 'blockGap' ) );
+	// Skip if gap value contains unsupported characters.
+	// Regex for CSS value borrowed from `safecss_filter_attr`, and used here
+	// because we only want to match against the value, not the CSS attribute.
+	$gap       = preg_match( '%[\\\(&=}]|/\*%', $gap ) ? null : $gap;
+	$id        = uniqid();
+	$class     = 'wp-block-gallery-' . $id;
+	$content   = preg_replace(
+		'/' . preg_quote( 'class="', '/' ) . '/',
+		'class="' . $class . ' ',
+		$content,
+		1
+	);
+	$gap_value = $gap ? $gap : 'var( --wp--style--block-gap, 0.5em )';
+	$style     = '.' . $class . '{ --wp--style--unstable-gallery-gap: ' . $gap_value . '}';
+	// Ideally styles should be loaded in the head, but blocks may be parsed
+	// after that, so loading in the footer for now.
+	// See https://core.trac.wordpress.org/ticket/53494.
+	add_action(
+		'wp_footer',
+		function () use ( $style ) {
+			echo '<style> ' . $style . '</style>';
+		}
+	);
+	return $content;
+}
 /**
  * Registers the `core/gallery` block on server.
  * This render callback needs to be here
```

Comment:
I know this comment is unrelated with this PR but I'd like to understand it a bit. It may hint to a bad conceptual thing in blocks?

Reply:
@ramonjd can you remember the background to this, it was added at the same time as the code to add the `data-id` attribute?

Reply:
@glendaviesnz sorry about the opaque comment there. 

This file was primarily added so that we can manipulate the image attributes by adding `data-id` in `block_core_gallery_data_id_backcompatibility()` See: https://github.com/WordPress/gutenberg/pull/35975

As for the addition of `register_block_core_gallery()`  full comment tells the story – the _what_, but not the _why_. 

```
 * This render callback needs to be here
 * so that the gallery styles are loaded in block-based themes.
 ```

Without this render function, block themes do not load their styles. For example, see this gallery:

With `register_block_type_from_metadata`

<img width="500" alt="Screen Shot 2022-02-10 at 10 16 22 am" src="https://user-images.githubusercontent.com/6458278/153307740-8a44521f-f609-4d82-a7a9-6954e5675222.png">


Without `register_block_type_from_metadata`

<img width="500" alt="Screen Shot 2022-02-10 at 10 16 44 am" src="https://user-images.githubusercontent.com/6458278/153307754-f071c6b1-2f1b-47c7-88e0-e183eaf0c33b.png">

These styles are not loaded, e.g., `.wp-block-gallery.has-nested-images` etc

<img width="800" alt="Screen Shot 2022-02-10 at 10 17 17 am" src="https://user-images.githubusercontent.com/6458278/153307810-a1ef0404-b271-476c-b1fd-43d2d4e300c8.png">

I'm not sure what's going on exactly though. I [investigated at the time](https://github.com/WordPress/gutenberg/pull/35975#issuecomment-960348378) and found that it affected the gallery only.

I'm still having trouble isolating the cause.

Reply:
Okay, so I've taken a better look and have cleaned things up a bit in https://github.com/WordPress/gutenberg/pull/38689

I'm still yet to work out why the existence of `register_block_type_from_metadata` in `gallery/index.php` is necessary in order for block themes to enqueue gallery block library styles.

Reply:
> I'm still yet to work out why the existence of register_block_type_from_metadata in gallery/index.php is necessary in order for block themes to enqueue gallery block library styles.

Seems to be because `should_load_separate_core_block_assets` filter returns `true` [where we're enqueuing block library styles](https://github.com/WordPress/gutenberg/blob/596afccfe1396584f11e7c5b7cc22bf1b24e4ced/lib/client-assets.php#L329).

Looks like that's occurring [here](https://github.com/wordpress/wordpress-develop/blob/8db1549c0f1af05fbb919ad2eeb7f7bff5a63da9/src/wp-includes/theme.php#L4192) for block themes.

Reply:
cc @aristath (see discussion above about styles not being loaded when they should)